### ISSUE TITLE
k3s versions >v1.27.3 ignore Tailscale IPv6 addresses on IPv4-only hosts, causing new fatal errors

### ISSUE THREAD

**Issue by lorenzo-w** [k3s-io/k3s#8459]
**Environmental Info:**
K3s Version: 
```
k3s version v1.27.6+k3s1 (bd04941a)\
go version go1.20.8
```

Node(s) CPU architecture, OS, and Version: 
- `Linux cloudlane-one-s0 5.15.0-84-generic #93-Ubuntu SMP Tue Sep 5 17:16:10 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux`
- Node only supports IPv4

Cluster Configuration:
3 servers with embedded etcd, 2 agents, affected node is an agent
```yaml
vpn-auth: "name=tailscale,joinKey=OMITTED"
node-external-ip: OMITTED
cluster-cidr: "10.42.0.0/16,2001:cafe:42:0::/56"
service-cidr: "10.43.0.0/16,2001:cafe:42:1::/112"
disable:
  - traefik
  - local-storage
etcd-expose-metrics: true
secrets-encryption: true
kube-apiserver-arg:
  - "oidc-issuer-url=OMITTED"
  - "oidc-client-id=OMITTED"
  - "oidc-groups-claim=groups"
kube-controller-manager-arg:
  - "bind-address=0.0.0.0"
  - "horizontal-pod-autoscaler-tolerance=0.3"
kube-proxy-arg:
  - "metrics-bind-address=0.0.0.0"
kube-scheduler-arg:
  - "bind-address=0.0.0.0"
tls-san: "k8s.cloudlane.one"
kubelet-arg:
  - "node-ip=0.0.0.0"
  - "eviction-hard=memory.available<100Mi,nodefs.available<5%"
  - "eviction-soft=memory.available<300Mi,nodefs.available<10%"
  - "eviction-soft-grace-period=memory.available=1m0s,nodefs.available=10m0s"
  - "eviction-max-pod-grace-period=120"
  - "system-reserved=cpu=200m,memory=400Mi,ephemeral-storage=8Gi"
  - "kube-reserved=cpu=0.3,memory=1.1Gi,ephemeral-storage=4Gi"
```

**Describe the bug:**
Dual stack IP did work on earlier versions of k3s v1.27.x via Tailscale VPN, even if the host system did not support IPv6. With newer versions it seems that k3s does not extract the IPv6 supplied by `tailscale status --json`, but only IPv4. I have several nodes, which don't support IPv6 but my cluster CIDR is dual stack, so updating these nodes renders them unusable.

**Steps To Reproduce:**

- Installed K3s v1.27.3+k3s1 on arbitrary server node + one or more agent-nodes without IPv4 support; with tailscale vpn and a dual stack cluster CIDR; installed via official installer script
- Verified that cluster is working as expected
- Upgraded to v1.27.6+k3s1

**Expected behavior:**
k3s-agent is able to start.

**Actual behavior:**
k3s-agent fails with a fatal error saying that the node IP, which is IPv4 only, is incompatible with dual stack cluster CIDR..

**Additional context / logs:**

Relevant error messages:
```
Sep 26 21:59:36 cloudlane-one-s0 k3s[149080]: time="2023-09-26T21:59:36Z" level=debug msg="Output from tailscale status --json: {\n  \"Version\": \"1.50.0-ta920f0231-geb5b0beea\",\n  \"TUN\": true,\n  \"BackendState\": \"Running\",\n  \"AuthURL\": \"\",\n  \"TailscaleIPs\": [\n    \"100.104.255.34\",\n    \"fd7a:115c:a1e0:ab12:4843:cd96:6268:ff22\"\n  ],\n ... REST OF LINE OMITTED
Sep 26 21:59:36 cloudlane-one-s0 k3s[149080]: time="2023-09-26T21:59:36Z" level=info msg="Node-ip changed to [100.104.255.34] due to VPN"
Sep 26 21:59:36 cloudlane-one-s0 k3s[149080]: time="2023-09-26T21:59:36Z" level=warning msg="VPN provider overrides node-external-ip parameter"
Sep 26 21:59:37 cloudlane-one-s0 k3s[149080]: time="2023-09-26T21:59:37Z" level=fatal msg="cluster-cidr: [10.42.0.0/16 2001:cafe:42::/56] and node-ip: [100.104.255.34], must share the same IP version (IPv4, IPv6 or dual-stack)"
```

As you can see the output of tailscale contains both IP versions, but k3s only picks the IPv4.

I have a suspicion that this commit could be responsible: https://github.com/k3s-io/k3s/commit/f21a01474d2c9c79cfc5e99d9eea4116458dccca
But I am far from fluent in Go, so this should be taken as a mere guess.

The error seems to not happen on host systems supporting IPv6.

**Comment by brandond** [comment]
@manuelbuil would you mind taking a look?

**Comment by brandond** [comment]
Because this node only has an ipv4 address, the ipv6 VPN IP is ignored. This was changed in https://github.com/k3s-io/k3s/pull/8077.

> Before this PR, as tailscale interface was always dualStack, we were always setting a dualstack node-ip. However, that created problems in ipv6-only env. This PR honors the original node-ip ip mode when setting tailscale's node-ip.

This was intended to fix ipv6-only nodes, but it sounds like we regressed ipv4-only nodes?

**Comment by VestigeJ** [comment]
TLDR - able to reproduce original behavior but the fix isn't perfect yet. The config.yaml has changed to try and see where the behavior in the cluster is coming from between these two variations.




##Environment Details
VERSION=v1.28.2+k3s1
COMMIT=

*Infrastructure*
- [X] Cloud


*Node(s) CPU architecture, OS, and version:*
```
Linux 5.14.21-150500.53-default x86_64 GNU/Linux
PRETTY_NAME="SUSE Linux Enterprise Server 15 SP5"
```

*Cluster Configuration:*

*Config.yaml:*
```
write-kubeconfig-mode: 644
debug: true
token: YOUR_TOKEN_HERE
selinux: true
node-external-ip: tailscale_IPV4,tailscale_IPV6 addr
protect-kernel-defaults: true
cluster-init: true
cluster-cidr: 10.42.0.0/16,2001:cafe:42:0::/56
service-cidr: 10.43.0.0/16,2001:cafe:42:1::/112
vpn-auth: "name=tailscale,joinKey=tskey-auth-KEYCONTCNTRL-sudosudosudo"

<details><summmary>Single Server Reproduction</summary>

<!-- Provide the command to install "k3s" -->

```
$ curl https://get.k3s.io/ --output install-"k3s".sh
$ sudo chmod +x install-"k3s".sh
$ sudo groupadd --system etcd && sudo useradd -s /sbin/nologin --system -g etcd etcd
$ sudo modprobe ip_vs_rr
$ sudo modprobe ip_vs_wrr
$ sudo modprobe ip_vs_sh
$ sudo printf "on_oovm.panic_on_oom=0 \nvm.overcommit_memory=1 \nkernel.panic=10 \nkernel.panic_ps=1 \nkernel.panic_on_oops=1 \n" > ~/90-kubelet.conf
$ sudo cp 90-kubelet.conf /etc/sysctl.d/
$ sudo systemctl restart systemd-sysctl
$ curl -fsSL https://tailscale.com/install.sh | sh
$ sudo INSTALL_K3S_VERSION=v1.28.2+k3s1 INSTALL_K3S_EXEC=server ./install-k3s.sh  //observe fatal failure prevents cluster from starting
$ sudo journalctl -u k3s --no-pager
```

**Results:**

$ sudo journalctl -u k3s --no-pager

```
time="2023-10-10T23:57:46Z" level=fatal msg="cluster-cidr: [10.42.0.0/16 2001:cafe:42::/56] and node-ip: [10.12.19.90], must share the same IP version (IPv4, IPv6 or dual-stack)"
```

</details>

<details><summmary><h4>Current behavior</h4></summary>

The nodes join when passing the kubelet-arg for all interfaces otherwise I end up needing to change the k3s.yaml config to target the tailscale IP to access the kubelet api with kubectl

*Config.yaml:*

```
write-kubeconfig-mode: 644
debug: true
token: YOUR_TOKEN_HERE
node-external-ip: 100.8.9.10
protect-kernel-defaults: true
cluster-init: true
cluster-cidr: 10.42.0.0/16,2001:cafe:42:0::/56
service-cidr: 10.43.0.0/16,2001:cafe:42:1::/112
vpn-auth: "name=tailscale,joinKey=tskey-auth-keykeykey"
node-ip: 100.8.9.10
kubelet-arg:
  - "node-ip=0.0.0.0"
```

Typical installation process as used above but targeting 

```
$ sudo INSTALL_K3S_COMMIT=ced25af5b1d591927860f1003abf41361ee15d34 INSTALL_K3S_EXEC=server ./install-k3s.sh
```

**Results:**

$ sudo journalctl -u k3s-agent --no-pager | grep -i fatal
```
Oct 11 18:24:42 ip-1-1-2-70 k3s[10206]: time="2023-10-11T18:24:42Z" level=fatal msg="unable to initialize Network Policy Controller: IPv6 was enabled but no IPv6 address was found on node"
```

$ kgn
```
NAME              STATUS   ROLES                       AGE     VERSION
ip-1-1-2-25       Ready    control-plane,etcd,master   4m24s   v1.28.2+k3s-ced25af5
ip-1-1-2-70       Ready    <none>                      80s     v1.28.2+k3s-ced25af5
```

$ kgp -A
```
NAMESPACE     NAME                                      READY   STATUS             RESTARTS      AGE
kube-system   coredns-6799fbcd5-8mjh8                   0/1     Running            2 (71s ago)   4m14s
kube-system   helm-install-traefik-5nrqh                1/1     Running            1 (51s ago)   4m14s
kube-system   helm-install-traefik-crd-nk6xg            1/1     Running            1 (51s ago)   4m14s
kube-system   local-path-provisioner-84db5d44d9-q2dzk   0/1     Error              4 (86s ago)   4m14s
kube-system   metrics-server-67c658944b-dghpf           0/1     CrashLoopBackOff   4 (7s ago)    4m14s
```

**Additional context / logs:**

I've got a lot of logs on the server node @rbrtbnfgl the only error on the agent is pasted above
</details>

**Comment by VestigeJ** [comment]
So follow up from yesterday with Roberto's help I've double checked the tailscale settings and I can confirm that the server node is exposing the subnet routes used by the cluster-cidr on both ipv4/6 but not the service-cidr. The k3s-agent is not exposing the subnet routes from the server and is not able to take in those config.yaml values for the node itself. 
 @manuelbuil Can you potentially comment on this?

**Comment by manuelbuil** [comment]
> the server node is exposing the subnet routes used by the cluster-cidr on both ipv4/6 but not the service-cidr

This is fine. Service-cidr does not need to be exposed because the kube-proxy load-balancing happens before going through the tailscale stack, so tailscale only see cluster-cidr IPs.

> The k3s-agent is not exposing the subnet routes from the server

That is bad. This was working before. @rbrtbnfgl @VestigeJ did you verify if this PR is creating this regression?

> is not able to take in those config.yaml values for the node itself

What do you mean by this?

**Comment by rbrtbnfgl** [comment]
It was working like this before the PR for me. I had to add the cluster CIDR on the tailscale web interface.

**Comment by manuelbuil** [comment]
@VestigeJ I have just built master and tried ipv4-only and dual-stack. Ping works in both IP Families and I can see tailscale routes correctly:

```
 azureuser@mbuil-vm1:~$ ip route show table 52
10.42.0.0/24 dev tailscale0 
100.67.128.152 dev tailscale0 
100.100.100.100 dev tailscale0 
azureuser@mbuil-vm1:~$ ip -6 route show table 52
2001:cafe:42::/64 dev tailscale0 metric 1024 pref medium
fd7a:115c:a1e0::/48 dev tailscale0 metric 1024 pref medium
```
```
azureuser@mbuil-vm0:~$ ip route show table 52
10.42.1.0/24 dev tailscale0 
100.72.166.115 dev tailscale0 
100.100.100.100 dev tailscale0 
azureuser@mbuil-vm0:~$ ip -6 route show table 52
2001:cafe:42:2::/64 dev tailscale0 metric 1024 pref medium
fd7a:115c:a1e0::/48 dev tailscale0 metric 1024 pref medium
```

**Comment by VestigeJ** [comment]
> > The k3s-agent is not exposing the subnet routes from the server
> 
> That is bad. This was working before. @rbrtbnfgl @VestigeJ did you verify if this PR is creating this regression?
>  I can re-run this with a previous release to vet the pre-existing behavior but yes I think historically it was working because that would've come up in the past on our tailscale validation.

> > is not able to take in those config.yaml values for the node itself
> 
> What do you mean by this?

When I say this I mean that the k3s-agent installation doesn't run at all if you try to pass these routes on the agent nodes configuration. So it wasn't a case of missing information on the agents side to permit those routes on tailscale.

**Comment by VestigeJ** [comment]
> It was working like this before the PR for me. I had to add the cluster CIDR on the tailscale web interface.

Interesting - I did not need to do additional configuration by using the joinKey token on the node but I would've expected the same behavior on both the agent and the server nodes.

**Comment by VestigeJ** [comment]
> @VestigeJ I have just built master and tried ipv4-only and dual-stack. Ping works in both IP Families and I can see tailscale routes correctly:
> 
> ```
>  azureuser@mbuil-vm1:~$ ip route show table 52
> 10.42.0.0/24 dev tailscale0 
> 100.67.128.152 dev tailscale0 
> 100.100.100.100 dev tailscale0 
> azureuser@mbuil-vm1:~$ ip -6 route show table 52
> 2001:cafe:42::/64 dev tailscale0 metric 1024 pref medium
> fd7a:115c:a1e0::/48 dev tailscale0 metric 1024 pref medium
> ```
> 
> ```
> azureuser@mbuil-vm0:~$ ip route show table 52
> 10.42.1.0/24 dev tailscale0 
> 100.72.166.115 dev tailscale0 
> 100.100.100.100 dev tailscale0 
> azureuser@mbuil-vm0:~$ ip -6 route show table 52
> 2001:cafe:42:2::/64 dev tailscale0 metric 1024 pref medium
> fd7a:115c:a1e0::/48 dev tailscale0 metric 1024 pref medium
> ```

I'll have to rebuild and try the latest commits - I think new ipv4/6 commits were added but I can't attest if those changes involved the code under test previously for tailscale. Manuel are your nodes Ubuntu nodes? My default is on SLES 15 SP5 so this might be something specific to that OS

**Comment by VestigeJ** [comment]
👍🏽 @rbrtbnfgl @manuelbuil 

Very different behavior on this latest COMMIT_ID cluster came up without issues



##Environment Details
COMMIT=3abc8b82ed0779ebaa5d0ca00165408ad085cc8f

*Infrastructure*
- [X] Cloud

*Node(s) CPU architecture, OS, and version:*

```
Linux 5.14.21-150500.53-default x86_64 GNU/Linux
PRETTY_NAME="SUSE Linux Enterprise Server 15 SP5"
```

*Cluster Configuration:*

```
NAME               STATUS   ROLES                       AGE   VERSION
ip-1-1-1-132       Ready    control-plane,etcd,master   24m   v1.28.2+k3s-3abc8b82
ip-1-1-2-164       Ready    control-plane,etcd,master   16m   v1.28.2+k3s-3abc8b82
ip-1-1-2-3         Ready    control-plane,etcd,master   16m   v1.28.2+k3s-3abc8b82
ip-1-1-3-49        Ready    <none>                      23m   v1.28.2+k3s-3abc8b82
```

*Config.yaml:*

```
write-kubeconfig-mode: 644
debug: true
token: YOUR_TOKEN_HERE
selinux: true
node-external-ip: 100.96.16.90,fd7a:iced:abe1:abe1:4843:cafe:6274:4b88
node-ip: 100.96.16.90,fd7a:iced:abe1:abe1:4843:cafe:6274:4b88
protect-kernel-defaults: true
cluster-init: true
cluster-cidr: 10.42.0.0/16,2001:cafe:42:0::/56
service-cidr: 10.43.0.0/16,2001:cafe:42:1::/112
vpn-auth: "name=tailscale,joinKey=tskey-auth-kqwerty-blookem"
```

<details><summmary><h4> YOUR_REPRODUCED_RESULTS_HERE </h4></summary>



```
$ curl https://get.k3s.io --output install-"k3s".sh
$ sudo chmod +x install-"k3s".sh
$ sudo groupadd --system etcd && sudo useradd -s /sbin/nologin --system -g etcd etcd
$ sudo modprobe ip_vs_rr
$ sudo modprobe ip_vs_wrr
$ sudo modprobe ip_vs_sh
$ sudo printf "on_oovm.panic_on_oom=0 \nvm.overcommit_memory=1 \nkernel.panic=10 \nkernel.panic_ps=1 \nkernel.panic_on_oops=1 \n" > ~/60-rke2-cis.conf or ~/90-kubelet.conf
$ sudo cp 60-rke2-cis.conf /etc/sysctl.d/ or 90-kubelet.conf
$ sudo systemctl restart systemd-sysctl
$ get_tailscale //curl installation of tailscale with systemd service
$ sudo tailscale up --accept-routes //start the systemd service with accepting new sub-routes 
$ ip a //get tailscale address for configuration of node-ip and node-external-ip 
$ conf //sudo vim /etc/rancher/k3s/config.yaml
$ sudo INSTALL_K3S_COMMIT=3abc8b82ed0779ebaa5d0ca00165408ad085cc8f INSTALL_K3S_EXEC=server ./install-k3s.sh
$ set_kubefig //export KUBECONFIG=/etc/rancher/k3s/k3s.yaml
$ kgn //kubectl get nodes
$ watch -n 3 kubectl get no,po,svc -A //observe primary server node as additional agents/servers join
$ ip route show table 52 //verify routing table is populated from tailscale
$ get_report //generate this template
```

**Results:**

table 52 on all nodes 

$ ip route show table 52
```
10.42.0.0/24 dev tailscale0
10.42.1.0/24 dev tailscale0
10.42.2.0/24 dev tailscale0
100.7.34.135 dev tailscale0
100.7.69.95  dev tailscale0
100.78.97.7  dev tailscale0
100.79.12.96 dev tailscale0
100.8.9.11   dev tailscale0
100.96.16.90 dev tailscale0
100.99.51.18 dev tailscale0
100.10.25.18 dev tailscale0
100.10.10.10 dev tailscale0
100.11.67.42 dev tailscale0
100.11.75.13 dev tailscale0
100.12.17.97 dev tailscale0
```
</details>
